### PR TITLE
Adding pagination to decrease the heap overhead size

### PR DESCRIPTION
**Pull request by MartinsIrbe**
- Calling Nitro with 30 PIDs at once.
- Paginating over Programmes, page by page (Page size = 30)

### CODE REVIEW

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java [hunk 1]
```diff
@@ -173,6 +209,16 @@ public String apply(Availability input) {
...
         });
     }
+
+    private List<Availability> getNitroAvailabilities(Clip clip) throws GlycerinException {
+        AvailabilityQuery query = AvailabilityQuery.builder()
+                .withDescendantsOf(clip.getPid())
+                .withPageSize(pageSize)
+                .build();
+
+        GlycerinResponse<Availability> availabilities = glycerin.execute(query);
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java]: This needs to do an exhaustive call.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java [hunk 2]
```diff
@@ -91,22 +107,42 @@ public GlycerinNitroContentAdapter(
         this.executor = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(60));
     }
 
+    public static final Function<Programme, Episode> toEpisode() {
+        return TO_EPISODE;
+    }
+
+    public static final Function<List<Programme>, List<Episode>> toEpisodesList() {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java]: wat.

Why are these needed?

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java [hunk 3]
```diff
@@ -75,6 +60,37 @@
...
+        }
+    };
+    private static final Predicate<Programme> IS_EPISODE = new Predicate<Programme>() {
+
+        @Override
+        public boolean apply(Programme input) {
+            return input.isEpisode();
+        }
+    };
+    private static final Function<List<Programme>, List<Programme>> IS_EPISODES_LIST = new Function<List<Programme>, List<Programme>>() {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java]: This is called `IS_EPISODES_LIST` which implies it's a predicate, while this clearly does a filter.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java [hunk 4]
```diff
@@ -171,18 +191,36 @@ public LocalOrRemoteNitroFetcher(ContentResolver resolver, NitroContentAdapter c
             }
         }
 
-        ImmutableSet<Item> fetched = contentAdapter.fetchEpisodes(toFetch);
-        return mergeItemsWithExisting(fetched, ImmutableSet.copyOf(Iterables.filter(resolvedItems.getAllResolvedResults(), Item.class)));
+        Iterator<List<Item>> fetchedItemListIterator = contentAdapter
+                .fetchEpisodes(toFetch)
+                .iterator();
+        ImmutableSet.Builder<Item> fetchedItemSet = ImmutableSet.builder();
+        while (fetchedItemListIterator.hasNext()) {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java]: Just use it as an `Iterable` in a for-each loop.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java [hunk 5]
```diff
@@ -171,18 +191,36 @@ public LocalOrRemoteNitroFetcher(ContentResolver resolver, NitroContentAdapter c
...
+            List<Item> itemList = fetchedItemListIterator.next();
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java]: There's an `.addAll` method for exactly these kinds of things.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java [hunk 6]
```diff
@@ -68,64 +74,83 @@ protected void runTask() {
...
-            reportStatus("Writing items");
-            writeContent(items, series, brands);
-        } catch (NitroException e) {
-            throw Throwables.propagate(e);
-        } catch (InterruptedException e) {
-            log.error("Could not lock item IDs", e);
-        } finally {
-            lock.unlock(episodeIds);
-            lock.unlock(containerIds);
+            Set<String> episodeIds = ImmutableSet.of();
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: You initialise these, and then overwrite them with a set. If you're not gonna use that value, just leave it uninitialised.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java [hunk 7]
```diff
@@ -68,64 +74,83 @@ protected void runTask() {
...
+
+                ResolveOrFetchResult<Item> resolvedItems = localOrRemoteFetcher
+                        .resolveItems(items);
+                ImmutableSet<Container> resolvedSeries = localOrRemoteFetcher
+                        .resolveOrFetchSeries(resolvedItems.getAll());
+                ImmutableSet<Container> resolvedBrands = localOrRemoteFetcher
+                        .resolveOrFetchBrand(resolvedItems.getAll());
+
+                Iterable<Series> series = Iterables.filter(
+                        Iterables.concat(resolvedSeries, resolvedBrands), Series.class
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: Why do you need to do this? Doesn't `.resolveOrFetchSeries()` above already return just `Series`?

Same for the brands thing below.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java [hunk 8]
```diff
@@ -41,6 +45,8 @@
     private final ContentWriter contentWriter;
     private final LocalOrRemoteNitroFetcher localOrRemoteFetcher;
     private final GroupLock<String> lock;
+    private int written = 0;
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: These shouldn't be fields. The task is only instantiated once and these will persist across invocations of `runTask`

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java [hunk 9]
```diff
@@ -68,64 +74,83 @@ protected void runTask() {
...
+                throw Throwables.propagate(e);
+            } catch (InterruptedException e) {
+                log.error("Could not lock item IDs", e);
+            } finally {
+                lock.unlock(episodeIds);
+                lock.unlock(containerIds);
+
+                reportStatus(String.format(
+                        "Written %d items of which %d failed",
+                        written,
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: You don't increase these counters anywhere.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedNitroItemSources.java [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+        @Override
+        public boolean hasNext() {
+            return episodesIterator.hasNext();
+        }
+
+        @Override
+        public List<Item> next() {
+            List<Episode> episodes = episodesIterator.next();
+            ImmutableList.Builder<Item> items = ImmutableList.builder();
+            currentEpisodesIterator = episodes.iterator();
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedNitroItemSources.java]: Again, just use it as an `Iterable`...

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedNitroItemSources.java [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+            } catch (GlycerinException e) {
+                throw Throwables.propagate(e);
+            }
+            return broadcasts;
+        }
+
+        private ImmutableListMultimap<String, Availability> getAvailabilities(List<Episode> episodes) {
+            ImmutableListMultimap<String, Availability> availabilities;
+            try {
+                availabilities = availabilities(episodes);
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedNitroItemSources.java]: Just do `return availabilities(episodes)` here. Same for the rest of them.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        public boolean hasNext() {
+            try {
+                if (currentResponse == null) { // Getting the first page.
+                    if (programmeQueries.hasNext()) {
+                        currentResponse = client.execute(programmeQueries.next());
+                        return currentResponse.hasNext();
+                    } else {
+                        return false;
+                    }
+                } else if (currentResponse.hasNext()) { // Getting the next page.
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java]: This comment is wrong.

Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java]: For that matter, I don't see you coping with pages of the response anywhere.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+                }
+            } catch (GlycerinException e) {
+                throw Throwables.propagate(e);
+            }
+        }
+
+        @Override
+        public List<Programme> next() {
+            try {
+                return ImmutableList.copyOf(currentResponse.getNext().getResults());
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java]: How will this ever return the current page of programmes?

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java [hunk 14]
```diff
@@ -151,6 +153,24 @@ public LocalOrRemoteNitroFetcher(ContentResolver resolver, NitroContentAdapter c
                 ImmutableSet.copyOf(Iterables.filter(resolvedItems.getAllResolvedResults(), Item.class)));
     }
 
+    public ResolveOrFetchResult<Item> resolveItemsForMerge(Iterable<List<Item>> items)
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java]: This is completely unnecessary. The default method already merges things.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java [hunk 15]
```diff
@@ -57,49 +59,35 @@ public PidUpdateController(
...
         try {
-            items = contentAdapter
+            Iterable<List<Item>> items = contentAdapter
                     .fetchEpisodes(ProgrammesQuery.builder()
                             .withPid(pid)
                             .withMixins(ANCESTOR_TITLES, CONTRIBUTIONS, IMAGES, GENRE_GROUPINGS)
                             .withPageSize(1)
                             .build());
-            items = localOrRemoteNitroFetcher.resolveItems(items).getAll();
+            mergedItems = localOrRemoteNitroFetcher.resolveItemsForMerge(items).getAll();
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: The new method on `LocalOrRemoteNitroFetcher` is pointless.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java [hunk 16]
```diff
@@ -113,27 +101,36 @@ public void updatePidFromNitro(HttpServletResponse response, @PathVariable("pid"
             return;
         }
 
-        if (items.isEmpty()) {
+        if (Iterables.getOnlyElement(items) == null) {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: Duplicated call for `.getOnlyElement`.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java [hunk 17]
```diff
@@ -113,27 +101,36 @@ public void updatePidFromNitro(HttpServletResponse response, @PathVariable("pid"
...
             log.error("No items found in Nitro for pid {}", pid);
             response.setStatus(HttpStatusCode.NOT_FOUND.code());
             response.setContentLength(0);
             return;
         }
 
-        Item item = Iterables.getOnlyElement(items);
+        List<Item> itemsList = Iterables.getOnlyElement(items);
+        Iterator<Item> itemIterator = itemsList.iterator();
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: Again, why do you need an `Iterator` rather than just using an `Iterable`?

**Comment by emilssolmanis** [comment]
Back to you @MartinsIrbe

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java (hunk 2 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java]: According to our style guide.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java (hunk 3 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java]: I will rename it :)

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java (hunk 13 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java]: you are right, we are skipping the first page.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java [hunk 18]
```diff
@@ -119,6 +121,31 @@ public String apply(@Nullable PidReference pidRef) {
...
+    public List<org.atlasapi.media.entity.Clip> clipsFor(PidReference ref) throws NitroException {
+        try {
+            Iterable<com.metabroadcast.atlas.glycerin.model.Clip> nitroClips = Iterables
+                    .transform(Iterables.filter(getNitroClips(ref), isClip), toClip);
+
+            if (Iterables.isEmpty(nitroClips)) {
+                log.warn("No programmes found for clipRefs {}", ref, new Function<PidReference, String>() {
+
+                    @Override
+                    public String apply(@Nullable PidReference pidRef) {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java]: Is this actually `Nullable` though? If it is, you're asking for a null pointer there...

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java [hunk 19]
```diff
@@ -173,6 +209,24 @@ public String apply(Availability input) {
...
+
+    private List<Availability> getNitroAvailabilities(Clip clip) throws GlycerinException {
+        ImmutableList.Builder<Availability> availabilities = ImmutableList.builder();
+
+        AvailabilityQuery query = AvailabilityQuery.builder()
+                .withDescendantsOf(clip.getPid())
+                .withPageSize(pageSize)
+                .build();
+
+        GlycerinResponse<Availability> availabilitiesResponse = glycerin.execute(query);
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java]: Why not just use the `.exhaust` method that's used everywhere?

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java [hunk 20]
```diff
@@ -75,6 +60,37 @@
...
+    private static final Function<Programme, Episode> TO_EPISODE = new Function<Programme, Episode>() {
+
+        @Override
+        public Episode apply(@Nullable Programme input) {
+            return input.getAsEpisode();
+        }
+    };
+    private static final Function<List<Programme>, List<Episode>> TO_EPISODES_LIST = new Function<List<Programme>, List<Episode>>() {
+
+        @Nullable
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java]: Assuming all these `Nullable` annotations were generated, since none of this code actually copes with nulls.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java (hunk 2 above)
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java]: Huh? Which point?

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java [hunk 21]
```diff
@@ -171,18 +173,31 @@ public LocalOrRemoteNitroFetcher(ContentResolver resolver, NitroContentAdapter c
             }
         }
 
-        ImmutableSet<Item> fetched = contentAdapter.fetchEpisodes(toFetch);
-        return mergeItemsWithExisting(fetched, ImmutableSet.copyOf(Iterables.filter(resolvedItems.getAllResolvedResults(), Item.class)));
+        Iterable<List<Item>> fetchedItemListIterator = contentAdapter.fetchEpisodes(toFetch);
+        ImmutableSet.Builder<Item> fetchedItemSet = ImmutableSet.builder();
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java]: You could just use `Iterables.concat` and make this 1 line.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java [hunk 22]
```diff
@@ -35,13 +39,33 @@
...
 
     public static final Logger log = LoggerFactory.getLogger(OffScheduleContentIngestTask.class);
+    public static final Function<Container, Series> TO_SERIES = new Function<Container, Series>() {
+
+        @Override
+        public Series apply(Container input) {
+            return (Series) input;
+        }
+    };
+    public static final Function<Container, Brand> TO_BRANDS = new Function<Container, Brand>() {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: Do these need to be public?

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java [hunk 23]
```diff
@@ -54,6 +78,14 @@ public OffScheduleContentIngestTask(
         this.pageSize = pageSize;
     }
 
+    public Function<Container, Series> toSeries() {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: Same, if the constants don't need to be public, neither do these.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java [hunk 24]
```diff
@@ -68,64 +100,83 @@ protected void runTask() {
...
 
-            episodeIds = ImmutableSet.copyOf(Iterables.transform(fetched,
-                    new Function<Item, String>() {
-                        @Override
-                        public String apply(Item input) {
-                            return BbcFeeds.pidFrom(input.getCanonicalUri());
-                        }
-                    }));
-            containerIds = topLevelContainerIds(fetched);
+        Iterator<List<Item>> itemIterator = fetched.iterator();
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: No need for an `Iterator`, just use a for-each loop.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java [hunk 25]
```diff
@@ -68,64 +100,83 @@ protected void runTask() {
...
+                lock.unlock(containerIds);
+
+                reportStatus(String.format(
+                        "Written %d items of which %d failed",
+                        written,
+                        failed
+                ));
+            }
         }
+        written = 0;
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: Probably safest to blat these out at the start of the job, not the end.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java (hunk 19 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java]: because it's used for `ProgrammeQuery`.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java (hunk 20 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroContentAdapter.java]: yes they were generated, I will remove them

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java (hunk 21 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/LocalOrRemoteNitroFetcher.java]: 👍

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java (hunk 22 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: they should be private as we give access to them via methods. :)

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java (hunk 23 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: 👍

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java (hunk 24 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/OffScheduleContentIngestTask.java]: 👍

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+         * This has been done to decrease the heap overhead when querying Nitro.
+         * @return returns true if the current Programme is present, else returns false.
+         */
+        @Override
+        public boolean hasNext() {
+            try {
+                if (currentResponse == null) { // Getting the first page.
+                    if (programmeQueries.hasNext()) {
+                        currentResponse = client.execute(programmeQueries.next());
+                        return true;
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedProgrammeRequest.java]: This isn't necessarily correct, the response might not have any results. This should instead be `return currentResponse.getResults().hasNext()` or something to that end.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java [hunk 27]
```diff
@@ -57,49 +60,38 @@ public PidUpdateController(
...
-            items = contentAdapter
+            Iterable<List<Item>> itemListIterable = contentAdapter
                     .fetchEpisodes(ProgrammesQuery.builder()
                             .withPid(pid)
                             .withMixins(ANCESTOR_TITLES, CONTRIBUTIONS, IMAGES, GENRE_GROUPINGS)
                             .withPageSize(1)
                             .build());
-            items = localOrRemoteNitroFetcher.resolveItems(items).getAll();
+
+            for (List<Item> itemList : itemListIterable) {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: This is clumsy, do something like
```
Iterable<Item> allItems = Iterables.concat(itemListIterable);
mergedItems = localOrRemote.resolve(allItems);
```

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java [hunk 28]
```diff
@@ -113,27 +105,33 @@ public void updatePidFromNitro(HttpServletResponse response, @PathVariable("pid"
             return;
         }
 
-        if (items.isEmpty()) {
+        List<Item> itemsList = Iterables.getOnlyElement(items);
+        if (itemsList == null) {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: This will never be true. `Iterables.getOnlyElement` will throw for an empty iterable, read the JavaDoc.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java [hunk 29]
```diff
@@ -113,27 +105,33 @@ public void updatePidFromNitro(HttpServletResponse response, @PathVariable("pid"
...
-        try {
-            contentWriter.createOrUpdate(item);
-            response.setStatus(HttpStatusCode.ACCEPTED.code());
-        } catch (IllegalArgumentException e) {
-            String message = String.format("Got more than 1 item from Nitro for pid %s", pid);
-            log.error(message);
-            response.setStatus(HttpStatusCode.SERVER_ERROR.code());
-            response.setContentLength(message.length());
-            response.getWriter().write(message);
+    private void createOrUpdateItems(HttpServletResponse response, @PathVariable("pid") String pid,
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: Having this annotated as a `@PathVariable` makes no sense.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java [hunk 30]
```diff
@@ -113,27 +105,33 @@ public void updatePidFromNitro(HttpServletResponse response, @PathVariable("pid"
             return;
         }
 
-        if (items.isEmpty()) {
+        List<Item> itemsList = Iterables.getOnlyElement(items);
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: For that matter, this needs flattening. Both of these endpoints operate on single items.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java [hunk 31]
```diff
@@ -113,27 +105,33 @@ public void updatePidFromNitro(HttpServletResponse response, @PathVariable("pid"
...
-            contentWriter.createOrUpdate(item);
-            response.setStatus(HttpStatusCode.ACCEPTED.code());
-        } catch (IllegalArgumentException e) {
-            String message = String.format("Got more than 1 item from Nitro for pid %s", pid);
-            log.error(message);
-            response.setStatus(HttpStatusCode.SERVER_ERROR.code());
-            response.setContentLength(message.length());
-            response.getWriter().write(message);
+    private void createOrUpdateItems(HttpServletResponse response, @PathVariable("pid") String pid,
+            Iterable<Item> itemIterator) throws IOException {
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: Since both endpoints operate on single items, having this as an `Iterable` makes no sense.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java (hunk 29 above)
Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PidUpdateController.java]: When using IntelliJ shortcut to create method this annotation is added by default. Removed.

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java (hunk 19 above)
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java]: So make one for `AvailabilityQuery`. This is inconsistent otherwise.

Comment (MartinsIrbe) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/GlycerinNitroClipsAdapter.java]: 👍

File: src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedNitroItemSources.java [hunk 32]
```diff
@@ -0,0 +1,323 @@
...
+                            "pc",
+                            "iptv-all",
+                            "captions")
+                    .build();
+
+            futures.add(executor.submit(exhaustingAvailabilityCallable(query)));
+
+
+            ListenableFuture<List<ImmutableList<Availability>>> all = Futures.allAsList(futures);
+            Iterable<Availability> list = null;
```
Comment (emilssolmanis) [inline src/main/java/org/atlasapi/remotesite/bbc/nitro/PaginatedNitroItemSources.java]: Leaving this uninitialised is much clearer and gets you nice compiler errors if someone messes with this in the future.
